From 55fd9d30b7e8b9c6403e86082ffb77f3706cdd31 Mon Sep 17 00:00:00 2001
From: "pj.chen" <chen.pj@inventec.com>
Date: Sun, 20 Jun 2021 08:17:24 +0000
Subject: [PATCH] [inventec][common][HwmonTempSensor] - Add
 totalThresholdNumber when HwmonTempSensor create

Symptom/Reason:
  For HwmonTempSensor, may have threshold config in one hwmon path, for example "tmp468".
  When setting the threshold, would also modify the property in entity-manager.
  So if we modify the sensor in the extention(temp2, temp3.....),
  would have error to find correct dbus interface.

Root Cause:
  In persistThreshold() in Thresholds.cpp, would find all threshold dbus interface.
  The threshold count default is the thresholds.size().
  But for the extention sensors for tmp468, all of the threshod are config in the temp1.
  And temp1 itself's thresholds only contain it's own. Whis is 4 for normal case.

  Here we need to find all threshold interface, so give the totalThresholdNumber.
  Which is 4*sensor numbers for normal case.

Solution/Change:
  Extend the input requirement for HwmonTempSensor creation.
  And give the label and total threshold size when setInitialProperties()

Entry Test:

THERMAL_Q89      | 23.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q90      | 24.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q91      | 26.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q92      | 26.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q93      | 29.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q94      | 28.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_U6       | 32.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 35.000    | 36.000    | na
THERMAL_U8       | 23.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 55.000    | 56.000    | na
THERMAL_U95      | 27.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na

ipmitool raw 0x04 0x26 0x61 0x01 0x06 0x00 0x00 0x00 0x00 0x00

THERMAL_Q89      | 23.000     | degrees C  | ok    | na        | 5.000     | 6.000     | 65.000    | 70.000    | na
THERMAL_Q90      | 24.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q91      | 26.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q92      | 26.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q93      | 28.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_Q94      | 28.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
THERMAL_U6       | 32.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 35.000    | 36.000    | na
THERMAL_U8       | 24.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 55.000    | 56.000    | na
THERMAL_U95      | 27.000     | degrees C  | ok    | na        | 5.000     | 10.000    | 65.000    | 70.000    | na
---
 include/HwmonTempSensor.hpp |  4 +++-
 src/HwmonTempMain.cpp       | 24 +++++++++++++++---------
 src/HwmonTempSensor.cpp     |  9 ++++++---
 3 files changed, 24 insertions(+), 13 deletions(-)

diff --git a/include/HwmonTempSensor.hpp b/include/HwmonTempSensor.hpp
index a2aed83..8376097 100644
--- a/include/HwmonTempSensor.hpp
+++ b/include/HwmonTempSensor.hpp
@@ -22,7 +22,8 @@ class HwmonTempSensor :
                     std::vector<thresholds::Threshold>&& thresholds,
                     std::vector<ipmidata::IpmiConfig>&& ipmiinfo,
                     const std::string& sensorConfiguration,
-                    const PowerState powerState);
+                    const PowerState powerState,
+                    size_t totalThresholdNumber);
     ~HwmonTempSensor();
     void setupRead(void);
 
@@ -33,6 +34,7 @@ class HwmonTempSensor :
     boost::asio::streambuf readBuf;
     std::string path;
     size_t errCount;
+    size_t totalThresholdNumber;
 
     void handleResponse(const boost::system::error_code& err);
     void checkThresholds(void) override;
diff --git a/src/HwmonTempMain.cpp b/src/HwmonTempMain.cpp
index e63d640..28e9700 100644
--- a/src/HwmonTempMain.cpp
+++ b/src/HwmonTempMain.cpp
@@ -205,20 +205,26 @@ void createSensors(
 
                 /* Parsing thresholds, for TMP468, need label in threshold config.*/
                 std::vector<thresholds::Threshold> sensorThresholds;
+                size_t totalThresholdNumber = 0;
 
-                if( !strncmp( sensorType, "xyz.openbmc_project.Configuration.TMP468" , strlen(sensorType))) {
+                if (!parseThresholdsFromConfig(*sensorData, sensorThresholds))
+                {
+                    std::cerr << "error populating thresholds for "
+                                << sensorName << "\n";
+                }
+                /* Counting all threshold config */
+                totalThresholdNumber = sensorThresholds.size();
+
+                /* Replace Threshold for TMP468*/
+                if( !strncmp( sensorType, "xyz.openbmc_project.Configuration.TMP468" , strlen(sensorType)))
+                {
+                    sensorThresholds.clear();
                     std::string thresholdLabel = sensorName.substr(0, sensorName.find(" "));
                     if (!parseThresholdsFromConfig(*sensorData, sensorThresholds,&thresholdLabel))
                     {
                         std::cerr << "error populating thresholds for "
                                   << sensorName << "\n";
                     }
-                } else {
-                    if (!parseThresholdsFromConfig(*sensorData, sensorThresholds))
-                    {
-                        std::cerr << "error populating thresholds for "
-                                  << sensorName << "\n";
-                    }
                 }
 
                 auto findPowerOn = baseConfiguration->second.find("PowerState");
@@ -247,7 +253,7 @@ void createSensors(
                     sensor = std::make_shared<HwmonTempSensor>(
                         *hwmonFile, sensorType, objectServer, dbusConnection,
                         io, sensorName, std::move(sensorThresholds), std::move(sensorIpmiConfig), 
-                        *interfacePath, readState);
+                        *interfacePath, readState, totalThresholdNumber);
                     sensor->setupRead();
                 }
                 // Looking for keys like "Name1" for temp2_input,
@@ -292,7 +298,7 @@ void createSensors(
                             dbusConnection, io, sensorName,
                             std::move(sensorThresholdsExtra),
                             std::move(sensorIpmiConfig),
-                            *interfacePath, readState);
+                            *interfacePath, readState, totalThresholdNumber);
                         sensor->setupRead();
                     }
                 }
diff --git a/src/HwmonTempSensor.cpp b/src/HwmonTempSensor.cpp
index 0a1856a..b5d7c30 100644
--- a/src/HwmonTempSensor.cpp
+++ b/src/HwmonTempSensor.cpp
@@ -46,12 +46,15 @@ HwmonTempSensor::HwmonTempSensor(
     boost::asio::io_service& io, const std::string& sensorName,
     std::vector<thresholds::Threshold>&& _thresholds,
     std::vector<ipmidata::IpmiConfig>&& _ipmiinfo,
-    const std::string& sensorConfiguration, const PowerState powerState) :
+    const std::string& sensorConfiguration,
+    const PowerState powerState,
+    size_t _totalThresholdNumber) :
     Sensor(boost::replace_all_copy(sensorName, " ", "_"),
            std::move(_thresholds), sensorConfiguration, objectType, maxReading,
            minReading, conn, powerState),
     std::enable_shared_from_this<HwmonTempSensor>(), objServer(objectServer),
-    inputDev(io, open(path.c_str(), O_RDONLY)), waitTimer(io), path(path)
+    inputDev(io, open(path.c_str(), O_RDONLY)), waitTimer(io), path(path),
+    totalThresholdNumber(_totalThresholdNumber)
 {
     sensorInterface = objectServer.add_interface(
         "/xyz/openbmc_project/sensors/temperature/" + name,
@@ -93,7 +96,7 @@ HwmonTempSensor::HwmonTempSensor(
     association = objectServer.add_interface(
         "/xyz/openbmc_project/sensors/temperature/" + name,
         association::interface);
-    setInitialProperties(conn);
+    setInitialProperties(conn, sensorName, totalThresholdNumber);
 }
 
 HwmonTempSensor::~HwmonTempSensor()
-- 
2.7.4

